R/autocompletion.R

```diff
@@ -1,26 +1,78 @@
-#' Extra polars auto completion
-#' @param activate bool default TRUE, enable chained auto-completion
-#' @name extra_auto_completion
-#' @return invisible NULL
-#' @noRd
+#' Polars code completion
 #'
-#' @details polars always supports auto completion via .DollarNames.
-#' However chained methods like x$a()$b()$? are not supported vi .DollarNames.
+#' @param activate Activate or deactivate the RStudio code completion
+#' @param mode One of `"auto"`, `"rstudio"`, or `"native"`. Automatic mode picks
+#' `"rstudio"` if `.Platform$GUI` is `"RStudio"`. `"native"` registers a custom
+#' line buffer completer with `utils:::rc.getOption("custom.completer")`.
+#' `"rstudio"` modifies RStudio code internal `.DollarNames` and function args
+#' completion, as the IDE does not behave well with
+#' `utils:::rc.getOption("custom.completer")`.
+#' @param verbose Print message of what mode is started.
 #'
-#' This feature experimental and not perfect. Any feedback is appreciated.
-#' Currently does not play that nice with Rstudio, as Rstudio backtick quotes any custom
-#' suggestions.
+#' @details
+#' Polars code completion has one implementation for an native terminal via
+#' `utils:::rc.getOption("custom.completer")` and one for Rstudio by intercepting
+#' Rstudio internal functions `.rs.getCompletionsFunction` &
+#' `.rs.getCompletionsDollar` in the loaded session environment `tools:rstudio`.
+#' Therefore, any error or slowness in the completion is likely to come from
+#' r-polars implementation.
+#'
+#' Either completers will evaluate the full line-buffer to decide what methods
+#' are available. Pressing tab will literally evaluate left-hand-side with any
+#' following side. This works swiftly for the polars lazy API, but it can take
+#' some time for the eager API depending on the size of the data and of the
+#' query.
+#'
+#' @return NULL
 #'
 #' @examples
-#' # auto completion via .DollarNames method
-#' e = pl$lit(42) # to autocomplete pl$lit(42) save to variable
-#' # then write `e$`  and press tab to see available methods
+#' \dontrun{
+#' # activate completion
+#' pl$code_completion()
+#'
+#  # method / property completion for chained expressions
+#  # add a $ and press tab to see methods of LazyFrame
+#' pl$LazyFrame(iris)
 #'
-#' # polars has experimental auto completion for chain of methods if all on the same line
-#' pl$extra_auto_completion() # first activate feature (this will 'annoy' the Rstudio auto-completer)
-#' pl$lit(42)$to_series() # add a $ and press tab 1-3 times
-#' pl$extra_auto_completion(activate = FALSE) # deactivate
-pl_extra_auto_completion = function(activate = TRUE) {
+#' # Arg + column-name completion
+#' # press tab inside group_by() to see args and/or column names.
+#' pl$LazyFrame(iris)$group_by()
+#'
+#' # deactivate like this or restart R session
+#' pl$code_completion(activate = FALSE)
+#' }
+pl_code_completion = function(
```

Comment:
Is there any reason to justify putting this under `pl` env?

Reply:
Not really, I didn't write this and didn't have a good reason to move it outside of `pl`. Would you prefer naming it `polars_completion()` instead?

Reply:
Hmmm, do you expect the user to execute this function directly?
If not, could a prefix like `.polars_` make sense?

Reply:
Yes the user can execute this directly to activate the custom completion since its opt-in for now

Reply:
I am not comfortable with the design of the function.
Why not make activation and deactivation separate functions, like `polars_code_completion_activate()` and `polars_code_completion_deactivate()`, respectively?

Reply:
Agreed on renaming. One question is how do we know which `mode` is used when we run `polars_deactivate_code_completion`? One solution would be to add a `completion` field in `polars_options()`, so that we (and the user) can always check if autocompletion is activated or not, and if so in which mode. Do you agree with this?

Reply:
Does the former mode matter when executing those functions?
I think we can display the information, perhaps via `polars_info()`?
(Similar to the one currently showing `thread_pool_size`)

Reply:
> Does the former mode matter when executing those functions?

Yes, it's not the same code to deactivate autocompletion depending on the mode

> I think we can display the information, perhaps via polars_info()?

We could display it there, but we still need to store the info somewhere so that `polars_info()` can retrieve it. That's why I thought we could store it as a `polars.autocompletion` option.

Reply:
Hmmm, I don't know if we can use option that way for preservation of the state.
Wouldn't you generally use environment type objects?

Reply:
I don't understand what you mean, here's what I have in mind for the value of `polars.autocompletion` option:

* default is autocompletion disabled, so it's `""` (or `NULL`)
* `polars_activate_code_completion(mode = "X")` =>  `X`
* `polars_activate_code_completion()` => retrieve the mode that was set by the user with `getOption("polars.autocompletion")`, deactivate autocompletion, and set the option to `""` (or `NULL`)

The option is reset when the session is restarted, but the autocompletion would be automatically deactivated so I don't think there's a problem here

Reply:
For now, in any case we require the user to run `polars_activate_code_completion()` so we don't have to take envvars into account

Reply:
The environment I am talking about is not an environment variable.
https://adv-r.hadley.nz/environments.html

Environment is commonly used in the R world to preserve state, right?
I don't think that options set by the user should be updated by the function.

Reply:
Oh right, I confused environments and envvars, I'll deal with that tomorrow or beginning of next week